File: internal/skirk/google_ip.go
package skirk

import (
	"errors"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

const defaultGoogleIPListFile = "assets/ip-list.txt"

type cachedListStore struct {
	mu         sync.RWMutex
	path       string
	entries    []string
	loadedPath string
	loadedAt   time.Time
	rotation   atomic.Uint64
}

var cache_list = &cachedListStore{path: defaultGoogleIPListPath()}

type googleIPCandidate struct {
	ip      string
	latency time.Duration
}

func defaultGoogleIP() string {
	return cache_list.Path()
}

func DefaultGoogleIP() string {
	return defaultGoogleIP()
}

func defaultGoogleIPListPath() string {
	if value := strings.TrimSpace(os.Getenv("SKIRK_GOOGLE_IP_LIST")); value != "" {
		return value
	}
	if value := strings.TrimSpace(os.Getenv("SKIRK_CACHED_LIST")); value != "" {
		return value
	}
	return defaultGoogleIPListFile
}

func (c *cachedListStore) Path() string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	if trimmed := strings.TrimSpace(c.path); trimmed != "" {
		return trimmed
	}
	return defaultGoogleIPListPath()
}

func (c *cachedListStore) SetPath(spec string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	trimmed := strings.TrimSpace(spec)
	if trimmed == "" {
		trimmed = defaultGoogleIPListPath()
	}
	if c.path != trimmed {
		c.path = trimmed
		c.entries = nil
		c.loadedPath = ""
		c.loadedAt = time.Time{}
		c.rotation.Store(0)
	}
}

func (c *cachedListStore) SetEntries(entries []string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.entries = normalizeGoogleIPEntries(entries)
	c.loadedPath = c.path
	c.loadedAt = time.Now().UTC()
	c.rotation.Store(0)
}

func (c *cachedListStore) Entries(spec string, limit int) ([]string, error) {
	ips, err := c.resolve(spec, limit)
	if err != nil {
		return nil, err
	}
	out := make([]string, len(ips))
	copy(out, ips)
	return out, nil
}

func resolveGoogleIPSpec(spec string) (string, error) {
	ips, err := resolveGoogleIPList(spec, 1)
	if err != nil {
		return "", err
	}
	if len(ips) == 0 {
		return "", errors.New("Google edge IP did not produce any IP")
	}
	return ips[0], nil
}

func resolveGoogleIPList(spec string, limit int) ([]string, error) {
	return cache_list.Entries(spec, limit)
}

func (c *cachedListStore) resolve(spec string, limit int) ([]string, error) {
	trimmed := strings.TrimSpace(spec)
	if trimmed == "" {
		trimmed = c.Path()
	}
	if limit <= 0 {
		limit = 12
	}

	if ip := net.ParseIP(trimmed); ip != nil {
		return []string{ip.String()}, nil
	}

	c.mu.RLock()
	cachedPath := c.loadedPath
	cachedEntries := append([]string(nil), c.entries...)
	c.mu.RUnlock()

	if cachedPath == trimmed && len(cachedEntries) > 0 {
		if limit < len(cachedEntries) {
			return append([]string(nil), cachedEntries[:limit]...), nil
		}
		return cachedEntries, nil
	}

	data, path, err := readGoogleIPList(trimmed)
	if err != nil {
		return nil, err
	}
	raw := normalizeGoogleIPEntries(parseGoogleIPsFromList(string(data)))
	if len(raw) == 0 {
		return nil, fmt.Errorf("google IP list %q did not contain a valid IP address", path)
	}

	ordered := prioritizeGoogleIPsByLatency(raw)

	c.mu.Lock()
	c.path = trimmed
	c.entries = append([]string(nil), ordered...)
	c.loadedPath = trimmed
	c.loadedAt = time.Now().UTC()
	c.mu.Unlock()

	if limit < len(ordered) {
		return append([]string(nil), ordered[:limit]...), nil
	}
	return ordered, nil
}

func normalizeGoogleIPEntries(entries []string) []string {
	seen := make(map[string]struct{}, len(entries))
	normalized := make([]string, 0, len(entries))
	for _, entry := range entries {
		ip := net.ParseIP(strings.TrimSpace(entry))
		if ip == nil {
			continue
		}
		canonical := ip.String()
		if _, ok := seen[canonical]; ok {
			continue
		}
		seen[canonical] = struct{}{}
		normalized = append(normalized, canonical)
	}
	return normalized
}

func parseGoogleIPsFromList(text string) []string {
	ips := make([]string, 0, 32)
	seen := make(map[string]struct{}, 32)
	for _, rawLine := range strings.Split(text, "\n") {
		line := strings.TrimSpace(rawLine)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if idx := strings.IndexByte(line, '#'); idx >= 0 {
			line = strings.TrimSpace(line[:idx])
			if line == "" {
				continue
			}
		}
		fields := strings.FieldsFunc(line, func(r rune) bool {
			switch r {
			case ',', ';', '	', ' ':
				return true
			default:
				return false
			}
		})
		candidates := append([]string(nil), fields...)
		candidates = append(candidates, strings.Trim(line, `"'`))
		for _, field := range candidates {
			field = strings.TrimSpace(strings.Trim(field, `"'`))
			if field == "" {
				continue
			}
			ip := net.ParseIP(field)
			if ip == nil {
				continue
			}
			canonical := ip.String()
			if _, ok := seen[canonical]; ok {
				continue
			}
			seen[canonical] = struct{}{}
			ips = append(ips, canonical)
		}
	}
	return ips
}

func prioritizeGoogleIPsByLatency(ips []string) []string {
	if len(ips) <= 1 {
		return ips
	}
	if strings.EqualFold(strings.TrimSpace(os.Getenv("SKIRK_GOOGLE_IP_PROBE_DISABLE")), "1") {
		return ips
	}
	probeTimeout := 450 * time.Millisecond
	if raw := strings.TrimSpace(os.Getenv("SKIRK_GOOGLE_IP_PROBE_TIMEOUT_MS")); raw != "" {
		if ms, err := time.ParseDuration(raw + "ms"); err == nil && ms >= 100*time.Millisecond && ms <= 5*time.Second {
			probeTimeout = ms
		}
	}
	probeCount := len(ips)
	if probeCount > 6 {
		probeCount = 6
	}
	candidates := make([]googleIPCandidate, 0, probeCount)
	for _, ip := range ips[:probeCount] {
		candidates = append(candidates, googleIPCandidate{
			ip:      ip,
			latency: measureGoogleIPLatency(ip, probeTimeout),
		})
	}
	sort.SliceStable(candidates, func(i, j int) bool {
		return candidates[i].latency < candidates[j].latency
	})
	ordered := make([]string, 0, len(candidates))
	for _, candidate := range candidates {
		ordered = append(ordered, candidate.ip)
	}
	for _, ip := range ips[probeCount:] {
		ordered = append(ordered, ip)
	}
	return ordered
}

func measureGoogleIPLatency(ip string, timeout time.Duration) time.Duration {
	start := time.Now()
	conn, err := net.DialTimeout("tcp", net.JoinHostPort(ip, "443"), timeout)
	if err != nil {
		return timeout + (10 * time.Second)
	}
	_ = conn.Close()
	latency := time.Since(start)
	if latency <= 0 {
		return time.Millisecond
	}
	return latency
}

func readGoogleIPList(spec string) ([]byte, string, error) {
	if data, err := os.ReadFile(spec); err == nil {
		return data, spec, nil
	}
	if filepath.IsAbs(spec) {
		return nil, spec, fmt.Errorf("could not read Google IP list %q", spec)
	}

	candidates := []string{spec}
	if cwd, err := os.Getwd(); err == nil {
		for dir := cwd; ; dir = filepath.Dir(dir) {
			candidates = append(candidates, filepath.Join(dir, spec))
			candidates = append(candidates, filepath.Join(dir, "resources", spec))
			parent := filepath.Dir(dir)
			if parent == dir {
				break
			}
		}
	}
	if exe, err := os.Executable(); err == nil {
		for dir := filepath.Dir(exe); ; dir = filepath.Dir(dir) {
			candidates = append(candidates, filepath.Join(dir, spec))
			candidates = append(candidates, filepath.Join(dir, "resources", spec))
			parent := filepath.Dir(dir)
			if parent == dir {
				break
			}
		}
	}

	seen := map[string]struct{}{}
	for _, candidate := range candidates {
		if _, ok := seen[candidate]; ok {
			continue
		}
		seen[candidate] = struct{}{}
		if data, err := os.ReadFile(candidate); err == nil {
			return data, candidate, nil
		}
	}
	return nil, spec, fmt.Errorf("could not read Google IP list %q", spec)
}

File: tools/google_ip.py
from __future__ import annotations

import ipaddress
import os
import socket
import threading
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DEFAULT_GOOGLE_IP_LIST = "assets/ip-list.txt"
LEGACY_CACHED_LIST_ENV = "SKIRK_CACHED_LIST"
GOOGLE_IP_LIST_ENV = "SKIRK_GOOGLE_IP_LIST"


class CachedList:
    def __init__(self, path: str | None = None) -> None:
        self._lock = threading.RLock()
        self._path = (path or os.environ.get(GOOGLE_IP_LIST_ENV) or os.environ.get(LEGACY_CACHED_LIST_ENV) or DEFAULT_GOOGLE_IP_LIST).strip() or DEFAULT_GOOGLE_IP_LIST
        self._entries: list[str] = []
        self._loaded_path: str = ""
        self._loaded_at: float = 0.0
        self._rotation = 0

    @property
    def path(self) -> str:
        with self._lock:
            return self._path

    def set_path(self, path: str | None) -> None:
        with self._lock:
            new_path = (path or os.environ.get(GOOGLE_IP_LIST_ENV) or os.environ.get(LEGACY_CACHED_LIST_ENV) or DEFAULT_GOOGLE_IP_LIST).strip() or DEFAULT_GOOGLE_IP_LIST
            if new_path != self._path:
                self._path = new_path
                self._entries = []
                self._loaded_path = ""
                self._loaded_at = 0.0
                self._rotation = 0

    def set_entries(self, entries: list[str]) -> None:
        with self._lock:
            self._entries = _normalize(entries)
            self._loaded_path = self._path
            self._loaded_at = time.time()
            self._rotation = 0

    def resolve(self, value: str | None, limit: int = 12) -> list[str]:
        spec = (value or self.path).strip() or self.path
        if limit <= 0:
            raise ValueError("limit must be greater than zero")
        try:
            return [str(ipaddress.ip_address(spec))]
        except ValueError:
            pass

        with self._lock:
            if self._loaded_path == spec and self._entries:
                return self._entries[:limit]

        candidates = _candidate_paths(spec)
        ips = _read_candidates(candidates)
        if not ips:
            raise ValueError(f"could not resolve Google IP from {spec!r}")
        ordered = _prioritize_by_latency(ips)
        with self._lock:
            self._entries = ordered
            self._loaded_path = spec
            self._loaded_at = time.time()
        return ordered[:limit]

    def resolve_one(self, value: str | None) -> str:
        return self.resolve(value, 1)[0]


cache_list = CachedList()


def _candidate_paths(spec: str) -> list[Path]:
    path = Path(spec)
    if path.exists():
        return [path]

    candidates = [ROOT / spec]
    cwd = Path.cwd()
    candidates.extend(parent / spec for parent in [cwd, *cwd.parents])
    candidates.extend(parent / spec for parent in [ROOT, *ROOT.parents])

    unique: list[Path] = []
    seen: set[str] = set()
    for candidate in candidates:
        key = str(candidate.resolve(strict=False))
        if key in seen:
            continue
        seen.add(key)
        unique.append(candidate)
    return unique


def _normalize(values: list[str]) -> list[str]:
    out: list[str] = []
    seen: set[str] = set()
    for value in values:
        try:
            ip = str(ipaddress.ip_address(value.strip()))
        except ValueError:
            continue
        if ip in seen:
            continue
        seen.add(ip)
        out.append(ip)
    return out


def _read_candidates(candidates: list[Path]) -> list[str]:
    ips: list[str] = []
    seen: set[str] = set()
    for candidate in candidates:
        try:
            text = candidate.read_text(encoding="utf-8")
        except OSError:
            continue
        for raw_line in text.splitlines():
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue
            if "#" in line:
                line = line.split("#", 1)[0].strip()
                if not line:
                    continue
            fields = [*line.replace(",", " ").replace(";", " ").split(), line]
            for field in fields:
                field = field.strip().strip("\"\'")
                if not field:
                    continue
                try:
                    ip = str(ipaddress.ip_address(field))
                except ValueError:
                    continue
                if ip in seen:
                    continue
                seen.add(ip)
                ips.append(ip)
        if ips:
            break
    return ips

def _tcp_connect_latency(ip: str, timeout: float = 0.45) -> float:
    started = time.perf_counter()
    try:
        with socket.create_connection((ip, 443), timeout=timeout):
            pass
    except OSError:
        return timeout + 10.0
    return max(time.perf_counter() - started, 0.0001)


def _prioritize_by_latency(ips: list[str]) -> list[str]:
    if len(ips) <= 1 or os.environ.get("SKIRK_GOOGLE_IP_PROBE_DISABLE", "").strip() == "1":
        return ips
    timeout = 0.45
    raw = os.environ.get("SKIRK_GOOGLE_IP_PROBE_TIMEOUT_MS", "").strip()
    if raw:
        try:
            ms = int(raw)
            if 100 <= ms <= 5000:
                timeout = ms / 1000.0
        except ValueError:
            pass
    probe_count = min(len(ips), 6)
    ranked = sorted(((ip, _tcp_connect_latency(ip, timeout)) for ip in ips[:probe_count]), key=lambda item: item[1])
    ordered = [ip for ip, _ in ranked]
    ordered.extend(ips[probe_count:])
    return ordered


def resolve_ip_list(value: str | None, limit: int = 12) -> list[str]:
    return cache_list.resolve(value, limit)


def resolve_ip(value: str | None) -> str:
    return cache_list.resolve_one(value)
